<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get VCF file with GWAS variants — get_vcf • gwasOncoX</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get VCF file with GWAS variants — get_vcf"><meta name="description" content="Downloads a tabix-indexed VCF file (.vcf.gz and .vcf.gz.tbi) with variants
associated with disease phenotypes (as found in the NHGRI-EBI GWAS catalog).
Only variants where the risk-allele is properly identified are included here.
The GWAS_HIT element of the INFO column has the following format:
rsid|risk_allele|pmid|tag_snp|p_valule|efo_id"><meta property="og:description" content="Downloads a tabix-indexed VCF file (.vcf.gz and .vcf.gz.tbi) with variants
associated with disease phenotypes (as found in the NHGRI-EBI GWAS catalog).
Only variants where the risk-allele is properly identified are included here.
The GWAS_HIT element of the INFO column has the following format:
rsid|risk_allele|pmid|tag_snp|p_valule|efo_id"><meta property="og:image" content="https://sigven.github.io/gwasOncoX/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-info" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">gwasOncoX</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sigven/gwasOncoX/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Get VCF file with GWAS variants</h1>
      <small class="dont-index">Source: <a href="https://github.com/sigven/gwasOncoX/blob/main/R/get_vcf.R" class="external-link"><code>R/get_vcf.R</code></a></small>
      <div class="d-none name"><code>get_vcf.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Downloads a tabix-indexed VCF file (<code>.vcf.gz</code> and <code>.vcf.gz.tbi</code>) with variants
associated with disease phenotypes (as found in the NHGRI-EBI GWAS catalog).
Only variants where the risk-allele is properly identified are included here.
The GWAS_HIT element of the INFO column has the following format:
rsid|risk_allele|pmid|tag_snp|p_valule|efo_id</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_vcf</span><span class="op">(</span>cache_dir <span class="op">=</span> <span class="cn">NA</span>, force_download <span class="op">=</span> <span class="cn">F</span>, build <span class="op">=</span> <span class="st">"grch37"</span>, cancer_only <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cache-dir">cache_dir<a class="anchor" aria-label="anchor" href="#arg-cache-dir"></a></dt>
<dd><p>Local directory for data download</p></dd>


<dt id="arg-force-download">force_download<a class="anchor" aria-label="anchor" href="#arg-force-download"></a></dt>
<dd><p>Logical indicating if local cache should be overwritten
(set to TRUE to force download if file exists in cache)</p></dd>


<dt id="arg-build">build<a class="anchor" aria-label="anchor" href="#arg-build"></a></dt>
<dd><p>genome assembly (grch37/grch38)</p></dd>


<dt id="arg-cancer-only">cancer_only<a class="anchor" aria-label="anchor" href="#arg-cancer-only"></a></dt>
<dd><p>logical indicating retrieval of all variants versus
cancer-associated variants only</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with file download information and metadata</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sigve Nakken.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

